File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/H.Большое_число/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>


bool comparator(const std::string &x, const std::string &y) {
    return (x + y) > (y + x);
}


void solution(int &n, std::vector<std::string>& nums) {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    */
    std::sort(nums.begin(), nums.end(), comparator);
    for (std::string& num : nums) {
        std::cout << num;
    }
    std::cout << "\n";
}

int main() {
    int n;
    std::cin >> n;
    std::vector<std::string> nums;
    for (int i = 0; i < n; ++i) {
        std::string num;
        std::cin >> num;
        nums.push_back(num);
    }
    solution(n, nums);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/H.Большое_число/solution.py
import sys
from functools import cmp_to_key

def comparator(x: str, y: int) -> bool:
    if x + y > y + x:
        return -1
    return 1

def solution(n: int, nums: list[str]) -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    """
    sys.stdout.write(''.join(sorted(nums, key=cmp_to_key(comparator))) + '\n')


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[str] = sys.stdin.readline().rstrip().split()
    solution(n, nums)
